Maximum Likelihood Demonstration

charlesknipp opened the pull request:
## Overview
I added a quick little MLE demonstration, which should work for _almost_ every AD backend (more on this later). This particular example was taken from [Kalman.jl](https://github.com/mschauer/Kalman.jl/blob/master/example/Kalman%20filter%20with%20automatic%20differentiation.ipynb) using Optimisers.jl to create a custom Newton's method.

## A Note on Automatic Differentiation
All the AD here is done via [DifferentiationInterface.jl](https://github.com/JuliaDiff/DifferentiationInterface.jl) to support something more universal. This allows us to quickly test various backends, switching to whichever suits the users' needs. Luckily, in this instance, every relevant backend can at least evaluate gradients. Unfortunately, since this is a Newton's method, the Hessian fails when using Enzyme.

## Second Order Differentiation
`Mooncake` is still working on their implementation, which is very exciting news; although that means it's functionality is limited to only basic gradient descent with appropriate hyperparametrs. 

`Enzyme` on the other hand, fully supports the calculation of the Hamilton vector product. In our demonstration however, this computation will fail with both DI and base Enzyme.

## Requests
- Make sure to `dev` both SSMProblems and GeneralisedFilters before resolving the toml
- If you have any suggestions, feel free to tear it to shreds

FredericWantiez commented on examples/kalman-filter/script.jl:
```diff
@@ -0,0 +1,98 @@
+    ps = [p]  # vector of filtered Gaussians
+    for i in 1:T
+        # Predict step
+        p = Φ * p ⊕ Gaussian(zero(z), Q)  # same as Gaussian(Φ*p.μ, Φ*p.Σ*Φ' + Q)
```
That's what I thought, the `particle` here is a `Gaussian` and we don't even need `Kalman`, the `GaussianDistributions` package has a `correct(::Gaussian)`

FredericWantiez commented:
Literate uses # to parse Markdown and ususally comments inside functions mess up everything

charlesknipp committed:
fixed type stability of linear filter

charlesknipp committed:
fixed type stability of linear filter

charlesknipp committed:
added MLE demonstration

charlesknipp committed:
added MLE demonstration

charlesknipp committed:
flipped sign of objective function

charlesknipp committed:
flipped sign of objective function

charlesknipp committed:
Merge remote-tracking branch 'origin/main' into ck/maximum-likelihood

charlesknipp committed:
reorganized and added Mooncake MWE

charlesknipp committed:
reorganized and added Mooncake MWE

shravanngoswamii commented:
> * I don't have the environment properly set up since GeneralisedFilters is not on the official registry, and I can't figure out how to add it to Project.toml

I guess, this may help you:
https://pkgdocs.julialang.org/dev/toml-files/#The-[sources]-section

Monorepo example is what we need:
```toml
[sources]
Example = {url = "https://github.com/JuliaLang/Example.jl", rev = "custom_branch"}
WithinMonorepo = {url = "https://github.org/author/BigProject", subdir = "SubPackage"}
SomeDependency = {path = "deps/SomeDependency.jl"}
```

gdalle commented:
Hi there! I'd be happy to help you figure out AD-related issues.
- Mooncake is not expected to handle second-order derivatives at the moment, we're working on it with @willtebbutt but it's a daunting task.
- Zygote's bug is rather disturbing, I can't really explain it.
- Enzyme even errors for the gradient when I try it on the latest version of everything. This makes it difficult for me to debug the Hessian. Since January, I have changed the `DI.gradient` bindings to forward as much as possible to Enzyme without meddling, so that might explain the difference wrt to @charlesknipp's first post. Could you maybe try it again to confirm that we're seeing the same things?

In general, when you encounter something you think is an issue with DI, I encourage you to open an issue on the DI repo with an MWE. It would help tremendously :)

charlesknipp commented:
> Enzyme even errors for the gradient when I try it on the latest version of everything. This makes it difficult for me to debug the Hessian. Since January, I have changed the `DI.gradient` bindings to forward as much as possible to Enzyme without meddling, so that might explain the difference wrt to @charlesknipp's first post. Could you maybe try it again to confirm that we're seeing the same things?

@gdalle ~I get an error for illegal type analysis with strict aliasing turned on. The computation succeeds with `Enzyme.API.strictAliasing!(false)`, which is good but suboptimal.~ it works with the [latest commit](https://github.com/TuringLang/SSMProblems.jl/pull/26/commits/25eb4d63a5b080d05eea224ce68dfe557b622a07#). 

Just to clarify, I'm working with DI version 0.6.43 and Enzyme version 0.13.30 on Julia 1.10.4 (my work computer)

EDIT: I'm not sure why, but forcing symmetry on the innovation covariance caused Enzyme to report a type instability. I converted it to a symmetric type instead, and somehow that fixed the inconsistency.

charlesknipp committed:
fixed KF type stability in Enzyme

charlesknipp committed:
fixed KF type stability in Enzyme

gdalle commented:
I'm afraid we'll have to concoct a pure-Enzyme MWE for the Hessian error, otherwise the Enzyme devs won't want to debug it

yebai commented:
> Second Order Differentiation

It is better to implement the basic ones first, i.e. first-order optimisation.

charlesknipp committed:
add MWE for Kalman filtering

charlesknipp committed:
add MWE for Kalman filtering

charlesknipp commented:
@yebai good call. I'll focus on the non-linear methods from now on. The resampling step is likely to cause some trouble. In some variational algorithms [here](https://gist.github.com/charlesknipp/9feac774e61af7c9f1ba4d3efd2e0155) I was able to use `Zygote.@ignore` to stop the gradient calculation consistent with the literature.

@gdalle I added a minimum working example, isolated from the SSMProblems/GeneralisedFilters interface. The gradient works fine, but the Hessian needs some work. I created some sugar akin to their demo from [here](https://enzyme.mit.edu/julia/stable/generated/autodiff/#Vector-forward-over-reverse), which clearly doesn't work. If you have some ideas let me know. I'm not sure how DI interfaces with Enzyme for the Hessian.

charlesknipp committed:
replaced second order optimizer and added backend testing

charlesknipp committed:
replaced second order optimizer and added backend testing

gdalle commented:
Your MWE looks simple enough to open an issue on Enzyme, so I'd say that's the right move

THargreaves committed:
Make HierarchicalSSM compatible with the BF (#70)

THargreaves committed:
Remove redundant computation from batch KF

THargreaves committed:
Factor out intermediate storage (#72)

THargreaves committed:
Remove prototype CUDA kernels from package

The batch Kalman filter was using two prototype custom CUDA kernels for performing fast Cholesky and regular-transposed matrix multiplication. The CUBLAS version of the former (technically LU) has been sped up now by better pointer creation in CUDA.jl. The second is still very slow and is the bottleneck to the computation. The prototype replacement is not fully tested though and even though it passed this packages unit tests, I found a case where it was giving unexplained incorrect values. For safety, it will be removed for now until the custom batch kernels are fully tested.

THargreaves committed:
Make RB particle creation more robust

log_weights gives a more reliable length than particles, which may be a strange type that we haven't defined an interface for yet.

THargreaves committed:
Export callback methods and types

charlesknipp committed:
Merge branch 'ck/maximum-likelihood' of https://github.com/TuringLang/AnalyticFilters.jl into ck/maximum-likelihood

charlesknipp committed:
fixed Mooncake errors for Bootstrap filter

charlesknipp commented:
Included in bulk with the guided filter
